src/KafkaFlow/Consumers/KafkaConsumerFlowManager.cs

```diff
@@ -0,0 +1,109 @@
+namespace KafkaFlow.Consumers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Confluent.Kafka;
+
+    internal class KafkaConsumerFlowManager : IKafkaConsumerFlowManager
+    {
+        private readonly IConsumer<byte[], byte[]> consumer;
+        private readonly CancellationToken stopCancellationToken;
+        private readonly ILogHandler logHandler;
+
+        private CancellationTokenSource heartbeatTokenSource;
+        private Task heartbeatTask;
+
+        private readonly List<TopicPartition> paused = new List<TopicPartition>();
+
+        public KafkaConsumerFlowManager(
+            IConsumer<byte[], byte[]> consumer,
+            CancellationToken stopCancellationToken,
+            ILogHandler logHandler)
+        {
+            this.consumer = consumer;
+            this.stopCancellationToken = stopCancellationToken;
+            this.logHandler = logHandler;
+        }
+
+        public void Pause(IReadOnlyCollection<TopicPartition> topicPartitions)
+        {
+            lock (this.paused)
+            {
+                topicPartitions = topicPartitions.Except(this.paused).ToList();
+
+                if (!topicPartitions.Any())
+                {
+                    return;
+                }
+
+                this.consumer.Pause(topicPartitions);
+                this.paused.AddRange(topicPartitions);
+
+                if (!this.AreAllPartitionsPaused())
+                {
+                    return;
+                }
+
+                this.heartbeatTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.stopCancellationToken);
+
+                this.heartbeatTask = Task.Run(
+                    () =>
+                    {
+                        try
+                        {
+                            this.consumer.Consume(this.heartbeatTokenSource.Token);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // Do nothing
+                        }
+                        catch (Exception ex)
+                        {
+                            this.logHandler.Error(
+                                "Error executing consumer heartbeat",
+                                ex,
+                                null);
+                        }
+                    },
+                    this.heartbeatTokenSource.Token);
+            }
+        }
+
+        public void Resume(IReadOnlyCollection<TopicPartition> topicPartitions)
+        {
+            lock (this.paused)
+            {
+                if (!topicPartitions.Any())
+                {
+                    return;
+                }
+
+                if (this.heartbeatTokenSource != null)
```

Comment:
I think it should validate if at least one partition was resumed before kill the heartbeat thread. 
Otherwise, if someone tried to resume invalid partitions the heartbeat thread will be cancelled (and it shouldn't).

Reply:
You are right, validation added